### PR TITLE
plan: covert max/min to Limit + Sort operators

### PR DESCRIPTION
**Pull request by shenli**
For SQL like `select max(c) from t;`, we could optimize it to `select c from t order by c desc limit 1;`.
For SQL like `select min(c) from t;`, we could optimize it to `select c from t order by c limit 1;`.

### CODE REVIEW

**Comment by shenli** [comment]
@hanfei1991 @winoros PTAL

File: plan/aggregation_eliminate.go [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+		}
+
+		// Add Sort and Limit operators.
+		// For max function, the sort order should be desc.
+		desc := f.GetName() == ast.AggFuncMax
+		// Compose Sort operator.
+		sort := Sort{}.init(a.ctx)
+		sort.ByItems = append(sort.ByItems, &ByItems{f.GetArgs()[0], desc})
+		sort.SetSchema(p.Schema().Clone())
+		sort.SetChildren(p.Children()...)
```
Comment (hanfei1991) [inline plan/aggregation_eliminate.go]: use setParentAndChildren

File: plan/aggregation_eliminate.go [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+			return p
+		}
+
+		// Add Sort and Limit operators.
+		// For max function, the sort order should be desc.
+		desc := f.GetName() == ast.AggFuncMax
+		// Compose Sort operator.
+		sort := Sort{}.init(a.ctx)
+		sort.ByItems = append(sort.ByItems, &ByItems{f.GetArgs()[0], desc})
+		sort.SetSchema(p.Schema().Clone())
```
Comment (hanfei1991) [inline plan/aggregation_eliminate.go]: the schema of sort should be set as its child's

File: plan/physical_plans.go [hunk 3]
```diff
@@ -562,6 +562,7 @@ func buildJoinSchema(joinType JoinType, join Plan, outerID int) *expression.Sche
 		return newSchema
 	}
 	return expression.MergeSchema(join.Children()[outerID].Schema(), join.Children()[1-outerID].Schema())
+
```
Comment (hanfei1991) [inline plan/physical_plans.go]: remove this

File: plan/aggregation_eliminate.go (hunk 2 above)
Comment (shenli) [reply plan/aggregation_eliminate.go]: So as the limit?

Comment (shenli) [reply plan/aggregation_eliminate.go]: @hanfei1991

**Comment by shenli** [comment]
@hanfei1991 @winoros PTAL

File: plan/logical_plan_builder.go [hunk 4]
```diff
@@ -63,14 +63,24 @@ func (p *LogicalAggregation) collectGroupByColumns() {
 func (b *planBuilder) buildAggregation(p LogicalPlan, aggFuncList []*ast.AggregateFuncExpr, gbyItems []expression.Expression) (LogicalPlan, map[int]int) {
 	b.optFlag = b.optFlag | flagBuildKeyInfo
 	b.optFlag = b.optFlag | flagAggregationOptimize
+	eliminateAgg := len(gbyItems) == 0
```
Comment (hanfei1991) [inline plan/logical_plan_builder.go]: This optimize is too complex. If we have an aggregate, we set flag eliminateAgg. That's enough.

**Comment by coocood** [comment]
@shenli 
This optimization should consider the index, the cost of sorting is much larger than aggregation. We should only do this when the column has an index.

**Comment by shenli** [comment]
@coocood The CBO framework should consider this.

File: plan/logical_plan_builder.go [hunk 5]
```diff
@@ -1491,6 +1505,9 @@ func (b *planBuilder) buildSelect(sel *ast.SelectStmt) LogicalPlan {
 		if b.err != nil {
 			return nil
 		}
+		if b.optFlag&flagAggEliminate == flagAggEliminate {
```
Comment (winoros) [inline plan/logical_plan_builder.go]: convert to topn will be no worse than original. it's ok not to do this

**Comment by shenli** [comment]
@hanfei1991 @winoros PTAL

**Comment by winoros** [comment]
So we can add some complex plan test.

**Comment by shenli** [comment]
@winoros Add a new test case. I will add more tests in tidb-test.

**Comment by hanfei1991** [comment]
@shenli plan test meets building fail.

**Comment by shenli** [comment]
@hanfei1991 Fixed.

**Pull request merged**